design/src/rfcs/rfc0024_better_constraint_violations.md

```diff
@@ -0,0 +1,509 @@
+        let regex = Self::compile_regex();
+
+        if regex.is_match(&string) {
+            Ok(string)
+        } else {
+            Err(crate::model::length_pattern_string::ConstraintViolation::Pattern(string))
+        }
+    }
+
+    pub fn compile_regex() -> &'static regex::Regex {
+        static REGEX: once_cell::sync::Lazy<regex::Regex> = once_cell::sync::Lazy::new(|| {
+            regex::Regex::new(r#"[a-f0-5]*"#).expect(r#"The regular expression [a-f0-5]* is not supported by the `regex` crate; feel free to file an issue under https://github.com/awslabs/smithy-rs/issues for support"#)
+        });
+
+        &REGEX
+    }
+}
+
+impl std::convert::TryFrom<std::string::String> for LengthPatternString {
+    type Error = crate::model::length_pattern_string::ConstraintViolation;
+
+    /// Constructs a `LengthPatternString` from an [`std::string::String`],
+    /// failing when the provided value does not satisfy the modeled constraints.
+    fn try_from(value: std::string::String) -> Result<Self, Self::Error> {
+        Self::check_length(&value)?;
+
+        let value = Self::check_pattern(value)?;
+
+        Ok(Self(value))
+    }
+}
+```
+
+Observe how a failure to adhere to the `@length` trait will short-circuit the
+evaluation of the constructor, when the value could technically also not adhere
+with the `@pattern` trait.
+
+Similarly, constrained structures fail upon encountering the first member that
+violates a constraint.
+
+Additionally, _in framework request deserialization code_:
+
+- collections whose members are constrained fail upon encountering the first
+  member that violates the constraint; and
+- maps whose keys and/or values are constrained fail upon encountering the
+  first violation; and
+
+In summary, any shape that is transitively constrained yields types whose
+constructors (both the internal one and the user-facing one) currently
+short-circuit upon encountering the first violation.
+
+### Solution proposal
+
+The deserializing architecture lends itself to be easily refactored so that we
+can collect constraint violations before returning them. Indeed, note that
+deserializers enforce constraint traits in a two-step phase: first, the
+_entirety_ of the unconstrained value is deserialized, _then_ constraint traits
+are enforced by feeding the entire value to the `TryFrom` constructor.
+
+We will introduce a `ConstraintViolations` type (note the plural) that
+represents a collection of constraint violations that can occur _within user
+application code_. Roughly:
+
+```rust
+pub ConstraintViolations<T>(pub(crate) Vec<T>);
+
+impl<T> IntoIterator<Item = T> for ConstraintViolations<T> { ... }
+
+impl std::convert::TryFrom<std::string::String> for LengthPatternString {
+    type Error = ConstraintViolations<crate::model::length_patterh_string::ConstraintViolation>;
+
+    fn try_from(value: std::string::String) -> Result<Self, Self::Error> {
+        // Check constraints and collect violations.
```

Comment:
We collect but we do not flatten. We collect `shape_name::ConstraintViolation`s or `shape_name::ConstraintViolationException`s. The latter ones are the ones that point to inner constraint violations, if any.

The implementation should not over-allocate. `Vec::new()` does not allocate. Only when we `push` to it we will.

Reply:
Supposing we had structure `((A, B), (C, D))` all with a constraints on them. If all constraints failed this would amount to something like `vec![vec![_, _], vec![_, _]]`? If we do flatten we'd get something like `vec![_, _, _, _]`? Both the flattened and unflattened approach three vectors. With the flattened vector requiring some extra work but being more cache friendly.

The alternative I'm thinking of is to only `Vec::new` once for the entire tree and instead push to a `&mut Vec<ValidationExceptionField>`.

Reply:
> The alternative I'm thinking of is to only Vec::new once for the entire tree and pass it to each conversion in the form of &mut Vec<ValidationExceptionField>.

Eagerly converting to `ValidationExceptionField`s has the downside that we'd be unable to map the constraint violation exceptions to other error shapes. Let's use [this thread](https://github.com/awslabs/smithy-rs/pull/2040#discussion_r1038169046).